typescript-test-samples/typescript-test-intro/list-buckets/tests/unit/list-buckets.test.ts

```diff
@@ -0,0 +1,158 @@
+ * subsets of tests easily.
+ */
+describe('list-buckets', () => {
+    describe('lambdaHandler()', () => {
+        let inputEvent: APIGatewayProxyEvent;
+
+        /**
+         * Shared test setup
+         *
+         * We're using a `beforeEach` block that runs setup prior to each unit test
+         * in this `describe` block. This lets us pull noise out of the test bodies
+         * and ensures that setup is consistent.
+         *
+         * The setup resets our s3Mock client back to its starting point between each
+         * unit test. This means we're starting from a clean state with each test.
+         *
+         * It also gets us a fresh instance of our inputEvent stub, with default values
+         * provided. We want to make sure that there is no data or context leaking
+         * between tests that could accidentally influence outcomes.
+         */
+        beforeEach(() => {
+            s3Mock.reset();
+            inputEvent = getEmptyAPIGatewayProxyEvent();
+        });
+
+        /**
+         * Unit tests.
+         *
+         * Now we come to the unit tests themselves. Each test checks a single
+         * scenario, with specific responses configured on the S3 client mock.
+         *
+         * The unit tests use the "Arrange, Act, Assert" pattern, setting up the
+         * data (Arrange), executing the handler function (Act), then checking that
+         * the result matches the inferface contract we expect (Assert).
+         */
+        it('returns success code and concatenated list for nonempty bucket list', async () => {
+            // Arrange
+            s3Mock.on(ListBucketsCommand).resolves({ Buckets: [{ Name: 'test1' }, { Name: 'test2' }] });
+
+            // Act
+            const result: APIGatewayProxyResult = await listBucketsLambdaHandler(inputEvent);
+
+            // Assert
+            expect(result.statusCode).toBe(200);
+            expect(result.body).toEqual('test1 | test2');
+        });
+        it('returns success code and empty list for empty bucket list', async () => {
+            s3Mock.on(ListBucketsCommand).resolves({});
+            const result: APIGatewayProxyResult = await listBucketsLambdaHandler(inputEvent);
+            expect(result.statusCode).toBe(200);
+            expect(result.body).toEqual('');
+        });
+        it('returns failure code and empty list for error getting bucket list', async () => {
+            s3Mock.on(ListBucketsCommand).rejects({ message: 'Failed to list buckets' });
+            const result: APIGatewayProxyResult = await listBucketsLambdaHandler(inputEvent);
+            expect(result.statusCode).toBe(500);
+            expect(result.body).toBe('');
+        });
+    });
+});
+
+/**
+ * Builds an empty API Gateway Proxy Event input object, with default values
+ * @returns {APIGatewayProxyEvent}
+ */
+function getEmptyAPIGatewayProxyEvent(): APIGatewayProxyEvent {
+    const result: APIGatewayProxyEvent = {
```

Comment:
In the Python project I had initially created a proxy event in a function. However later we refactored to put this event into the events folder as a JSON object. (see this: https://github.com/aws-samples/serverless-test-samples/blob/fad4659150c06ca0acc4b63d28c3be0a1a30729a/python-test-samples/apigw-lambda/tests/unit/mock_test.py#L14). I think the JSON approach is cleaner. What do you think?

Reply:
Did it this way to keep the TypeScript type engine/compiler happy. I'll see if we can do a clean read+cast off of a JSON file to get the best of both worlds.

Reply:
Also concerned about node imports pulling the JSON struct in as a singleton, then having the unit tests clobber each other. The way we have it now, its a Factory, so each test gets its own event instance. Will investigate this more deeply too.